src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -660,98 +661,118 @@ public long GetNumberOfCounterEntries(DocumentsOperationContext context)
             return fst.NumberOfEntries;
         }
 
-        public void UpdateDocumentCounters(DocumentsOperationContext context, BlittableJsonReaderObject doc, string docId, BlittableJsonReaderObject metadata, List<CounterOperation> countersOperations)
+        public void UpdateDocumentCounters(DocumentsOperationContext context, BlittableJsonReaderObject doc, string docId, List<CounterOperation> countersOperations)
         {
-            List<string> updates = null;
-            if (metadata.TryGet(Constants.Documents.Metadata.Counters, out BlittableJsonReaderArray counters))
+            BlittableJsonReaderArray metadataCounters = null;
+            var initCounterListSize = countersOperations.Count;
+            if (doc.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) &&
+                metadata.TryGet(Constants.Documents.Metadata.Counters, out metadataCounters))
             {
-                foreach (var operation in countersOperations)
-                {
-                    // we need to check the updates to avoid inserting duplicate counter names
-                    int loc = updates?.BinarySearch(operation.CounterName, StringComparer.OrdinalIgnoreCase) ??
-                              counters.BinarySearch(operation.CounterName, StringComparison.OrdinalIgnoreCase);
-
-                    switch (operation.Type)
-                    {
-                        case CounterOperationType.Increment:
-                        case CounterOperationType.Put:
-                            if (loc < 0)
-                            {
-                                CreateUpdatesIfNeeded();
-                                updates.Insert(~loc, operation.CounterName);
-                            }
-
-                            break;
-                        case CounterOperationType.Delete:
-                            if (loc >= 0)
-                            {
-                                CreateUpdatesIfNeeded();
-                                updates.RemoveAt(loc);
-                            }
-                            break;
-                        case CounterOperationType.None:
-                        case CounterOperationType.Get:
-                            break;
-                        default:
-                            throw new ArgumentOutOfRangeException($"Unknown value {operation.Type}");
-                    }
-                }
+                initCounterListSize += metadataCounters.Length;
             }
-            else
+
+            //Todo Maybe should not read from the table if there is metadata on document
```

Comment:
Maybe should not read from the table if there is metadata on the document?
The problem has been rise because there was no metadata in the document.
But if there is, we can avoid reading from the counter table and rely on the metadata on the document like before

Reply:
https://github.com/ravendb/ravendb/pull/7550#discussion_r215441220